client/my-sites/upgrades/domain-search/site-redirect.jsx

```diff
@@ -1,63 +1,84 @@
-	observe = require( 'lib/mixins/data-observe' );
+import HeaderCake from 'components/header-cake';
+import Main from 'components/main';
+import SiteRedirectStep from './site-redirect-step';
+import isSiteUpgradeable from 'state/selectors/is-site-upgradeable';
+import { getSelectedSite, getSelectedSiteId, getSelectedSiteSlug } from 'state/ui/selectors';
+import QueryProductsList from 'components/data/query-products-list';
 
-var SiteRedirect = React.createClass( {
-	mixins: [ observe( 'productsList', 'sites' ) ],
-
-	propTypes: {
+class SiteRedirect extends Component {
+	static propTypes = {
+		cart: React.PropTypes.object.isRequired,
+		selectedSite: React.PropTypes.object,
+		selectedSiteSlug: React.PropTypes.string,
+		isSiteUpgradeable: React.PropTypes.bool,
 		productsList: React.PropTypes.object.isRequired,
-		sites: React.PropTypes.object.isRequired
-	},
-
-	componentWillMount: function() {
-		this.checkSiteIsUpgradeable();
-	},
+		translate: React.PropTypes.func.isRequired,
+	};
 
-	componentDidMount: function() {
-		this.props.sites.on( 'change', this.checkSiteIsUpgradeable );
-	},
+	constructor() {
+		super();
+		this.handleBackToDomainSearch = this.handleBackToDomainSearch.bind( this );
+	}
 
-	componentWillUnmount: function() {
-		this.props.sites.off( 'change', this.checkSiteIsUpgradeable );
-	},
+	componentDidMount() {
+		this.checkSiteIsUpgradeable();
+	}
 
-	checkSiteIsUpgradeable: function( ) {
-		var selectedSite = this.props.sites.getSelectedSite();
+	componentWillReceiveProps() {
+		this.checkSiteIsUpgradeable();
+	}
 
-		if ( selectedSite && ! selectedSite.isUpgradeable() ) {
+	checkSiteIsUpgradeable() {
+		if ( this.props.selectedSite && ! this.props.isSiteUpgradeable ) {
 			page.redirect( '/domains/add' );
 		}
-	},
+	}
 
-	backToDomainSearch: function() {
-		page( '/domains/add/' + this.props.sites.getSelectedSite().slug );
-	},
+	handleBackToDomainSearch() {
+		page( '/domains/add/' + this.props.selectedSiteSlug );
+	}
+
+	render() {
+		const {
+			cart,
+			selectedSite,
+			productsList,
+			translate,
+		} = this.props;
 
-	render: function() {
 		return (
 			<Main>
-				<HeaderCake onClick={ this.backToDomainSearch }>
-					{ this.translate( 'Redirect a Site' ) }
+				<HeaderCake onClick={ this.handleBackToDomainSearch }>
+					{ translate( 'Redirect a Site' ) }
 				</HeaderCake>
 
 				<SiteRedirectStep
-					cart={ this.props.cart }
-					products={ this.props.productsList.get() }
-					selectedSite={ this.props.sites.getSelectedSite() } />
+					cart={ cart }
+					products={ productsList }
+					selectedSite={ selectedSite } />
+
+				<QueryProductsList />
 			</Main>
 		);
 	}
-} );
+}
 
-module.exports = SiteRedirect;
+export default connect(
+	( state ) => ( {
+		selectedSite: getSelectedSite( state ),
+		selectedSiteSlug: getSelectedSiteSlug( state ),
+		isSiteUpgradeable: isSiteUpgradeable( state, getSelectedSiteId( state ) ),
+		productsList: state.productsList.items,
```

Comment:
I don't think we ever need all products, that's why I would think twice before we introduce selector just to have it. This is a good indicator that it needs further refactoring :)